src/db/mysql/models.rs

```diff
@@ -0,0 +1,352 @@
+
+no_arg_sql_function!(last_insert_id, Integer);
+
+/// Run the diesel embedded migrations
+///
+/// Mysql DDL statements implicitly commit which could disrupt MysqlPool's
+/// begin_test_transaction during tests. So this runs on its own separate conn.
+pub fn run_embedded_migrations(settings: &Settings) -> Result<()> {
+    let conn = MysqlConnection::establish(&settings.database_url).unwrap();
+    Ok(embedded_migrations::run(&conn)?)
+}
+
+pub struct MysqlDbPool {
+    pool: Pool<ConnectionManager<MysqlConnection>>,
+}
+
+// XXX: to become a db::DbPool trait
+impl MysqlDbPool {
+    pub fn new(settings: &Settings) -> Result<Self> {
+        let manager = ConnectionManager::<MysqlConnection>::new(settings.database_url.as_ref());
+        let mut builder = Pool::builder().max_size(settings.database_pool_max_size.unwrap_or(10));
+        #[cfg(test)]
+        {
+            if settings.database_use_test_transactions {
+                builder = builder.connection_customizer(Box::new(TestTransactionCustomizer));
+            }
+        }
+        Ok(Self {
+            pool: builder.build(manager)?,
+        })
+    }
+
+    pub fn get(&self) -> Result<MysqlDb> {
+        Ok(MysqlDb {
+            #[cfg(not(test))]
+            conn: self.pool.get()?,
+            #[cfg(test)]
+            conn: LoggingConnection::new(self.pool.get()?),
+        })
+    }
+}
+
+pub struct MysqlDb {
+    #[cfg(not(test))]
+    pub(super) conn: PooledConnection<ConnectionManager<MysqlConnection>>,
+    #[cfg(test)]
+    pub(super) conn: LoggingConnection<PooledConnection<ConnectionManager<MysqlConnection>>>,
+}
+
+impl MysqlDb {
+    pub fn get_collection_id_sync(
+        &self,
+        params: &params::GetCollectionId,
+    ) -> Result<results::GetCollectionId> {
+        let id = if let Some(id) = get_std_collection_id(params) {
+            id
+        } else {
+            sql_query("SELECT id FROM collections WHERE name = ?")
+                .bind::<Text, _>(params)
+                .get_result::<IdResult>(&self.conn)?
+                .id
+        };
+        Ok(id)
+    }
+
+    pub fn get_collections_sync(
+        &self,
+        params: &params::GetCollections,
+    ) -> Result<results::GetCollections> {
+        sql_query("SELECT id, modified FROM collections WHERE user_id = ?")
+            .bind::<Integer, _>(params.user_id as i32)
+            .load::<CollectionsResult>(&self.conn)?
+            .into_iter()
+            .map(|cr| {
+                self.get_collection_name(cr.id)
+                    .map(|name| (name, cr.modified))
+            })
+            .collect()
+    }
+
+    pub fn create_collection_sync(&self, name: &str) -> Result<i32> {
+        let collection_id = self.conn.transaction(|| {
+            sql_query("INSERT INTO collections (name, modified) VALUES (?, ?)")
+                .bind::<Text, _>(name)
+                .bind::<BigInt, _>(ms_since_epoch())
+                .execute(&self.conn)?;
```

Comment:
This might fail due to two concurrent attempts to insert the same collection; you could automatically recover in that case by detecting the uniqueness constraint violation and reading the inserted row, similar to https://github.com/mozilla-services/server-syncstorage/blob/master/syncstorage/storage/sql/__init__.py#L921

Reply:
added a TODO